mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -1710,6 +1865,419 @@ public override string ToString ()
+			fixed (byte* ptr = data)
+				if (!NativeConvert.TryCopy (address, (IntPtr) ptr))
+					throw new ArgumentException ("address", "Failed to convert to native struct");
+			data_len = size;
+			sa_family = address.sa_family;
+		}
+
+		public unsafe void CopyTo (Sockaddr address)
+		{
+			if (address == null)
+				throw new ArgumentNullException ("address");
+			if (data_len < 0 || data_len > data.Length)
+				throw new ArgumentException ("this", "data_len < 0 || data_len > data.Length");
+
+			fixed (byte* ptr = data)
+				if (!NativeConvert.TryCopy ((IntPtr) ptr, data_len, address))
+					throw new ArgumentException ("this", "Failed to convert from native struct");
+		}
+
+		public override string ToString ()
+		{
+			var sb = new StringBuilder ();
+			sb.AppendFormat ("{{sa_family={0}, data_len={1}, data=(", sa_family, data_len);
+			for (int i = 0; i < data_len; i++) {
+				if (i != 0)
+					sb.Append (" ");
+				sb.Append (data[i].ToString ("x2"));
+			}
+			sb.Append (")");
+			return sb.ToString ();
+		}
+
+		public override int GetHashCode ()
+		{
+			unchecked {
+				int hash = 0x1234;
+				for (int i = 0; i < data_len; i++)
+					hash += i ^ data[i];
+				return hash;
+			}
+		}
+		public override bool Equals (object obj)
+		{
+			if (!(obj is SockaddrStorage))
+				return false;
+			return Equals ((SockaddrStorage) obj);
+		}
+		public bool Equals (SockaddrStorage value)
+		{
+			if (value == null)
+				return false;
+			if (data_len != value.data_len)
+				return false;
+			for (int i = 0; i < data_len; i++)
+				if (data[i] != value.data[i])
+					return false;
+			return true;
+		}
+	}
+
+	[CLSCompliant (false)]
+	public sealed class SockaddrUn : Sockaddr, IEquatable<SockaddrUn> {
+		public UnixAddressFamily sun_family { // AF_UNIX
+			get { return sa_family; }
+			set { sa_family = value; }
+		}
+		public byte[] sun_path;
+		public long sun_path_len; // Indicates how many bytes of sun_path are valid. Must not be larger than sun_path.Length.
+
+		internal override byte[] DynamicData ()
+		{
+			return sun_path;
+		}
+		internal override long GetDynamicLength ()
+		{
+			return sun_path_len;
+		}
+		internal override void SetDynamicLength (long value)
+		{
+			sun_path_len = value;
+		}
+
+		[DllImport (Syscall.MPH, SetLastError=true,
+				EntryPoint="Mono_Posix_SockaddrUn_get_sizeof_sun_path")]
+		static extern int get_sizeof_sun_path ();
+		static readonly int sizeof_sun_path = get_sizeof_sun_path ();
+
+		public SockaddrUn () : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			sun_path = new byte[sizeof_sun_path];
+			sun_path_len = 0;
+		}
+		public SockaddrUn (int size) : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			sun_path = new byte[size];
+			sun_path_len = 0;
+		}
+		public SockaddrUn (string path, bool @abstract = false) : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
```

Comment:
> Naming-wise, I suppose abstractNamespace might be better?

I've renamed it (and IsAbstract) to linuxAbstractNamespace / IsLinuxAbstractNamespace (which should make clearer what it is and that it is linux-specific).

> 
> Then there's semantics: does SockaddrUn need to know about this?
> 
> On the one hand, it kinda does, because UnixEncoding.Instance.GetBytes("\u0000value") skips the leading \u0000, which is arguably a bug. If we fix that bug, then SockaddrUn no longer needs to care about these abstract namespaces; the caller would just use a NUL-prefixed string, as is done on Linux.

The problem is that UnixEncoding treats '\0' as the magical char to include a non-UTF8-byte. Currently UnixEncoding will treat "\0v" as "add the byte (byte)'v'". (IMHO it should throw an exception in this case because this means there is no 1:1 mapping between string and byte array, but that's not really important for this pull request.) For "\0v" we could change the behavior of UnixEncoding, but e.g. for "\0\xff" we can't (because that the normal representation of the byte sequence {0xff}).

We can say that the user should use "\0\0value" to denote an abstract namespace (the first '\0' will then tell UnixEncoding that the second '\0' should be output as (byte)'\0'), but IMHO that's confusing, and it would be better to keep the current version with a separate flag. The entire idea of UnixEncoding is based on the fact that strings in Unix never contain a 0-byte, and I think this might be the only point in the Linux API where something which normally considered a string can contain a 0-byte, so I think it's justified to treat this as a special case.